src/ucp/core/ucp_mm.c

```diff
@@ -178,24 +178,102 @@ static ucs_status_t ucp_mem_alloc(ucp_context_h context, size_t length,
+        }
+
+        if (params->flags & UCP_MEM_MAP_FIXED) {
+            flags |= UCT_MD_MEM_FLAG_FIXED;
+        }
+    }
+
+    return flags;
+}
+
+/* Matrix of behavior
+ * |-----------------------------------------------------------------------------|
+ * | parameter |                          value                                  |
+ * |-----------|-----------------------------------------------------------------|
+ * | ALLOCATE  |  0  |     1     |  0  |  0  |  1  |     1     |  0  |     1     |
+ * | FIXED     |  0  |     0     |  1  |  0  |  1  |     0     |  1  |     1     |
+ * | addr      |  0  |     0     |  0  |  1  |  0  |     1     |  1  |     1     |
+ * |-----------|-----|-----------|-----|-----|-----|-----------|-----|-----------|
+ * | result    | err | alloc/reg | err | reg | err | alloc/reg | err | alloc/reg |
+ * |           |     |           |     |     |     |  (hint)   |     | (fixed)   |
+ * |-----------------------------------------------------------------------------|
+ */
+static inline ucs_status_t ucp_mem_map_check_and_adjust_params(ucp_mem_map_params_t *params)
+{
+    if (!(params->field_mask & UCP_MEM_MAP_PARAM_FIELD_LENGTH)) {
+        ucs_error("The length value for mapping memory isn't set: %s",
+                  ucs_status_string(UCS_ERR_INVALID_PARAM));
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    /* First of all, define all fields */
+    if (!(params->field_mask & UCP_MEM_MAP_PARAM_FIELD_ADDRESS)) {
+        params->field_mask |= UCP_MEM_MAP_PARAM_FIELD_ADDRESS;
+        params->address = NULL;
+    }
+
+    if (!(params->field_mask & UCP_MEM_MAP_PARAM_FIELD_FLAGS)) {
+        params->field_mask |= UCP_MEM_MAP_PARAM_FIELD_FLAGS;
+        params->flags = 0;
+    }
+
+    if ((params->flags & UCP_MEM_MAP_FIXED) &&
+        (!params->address ||
+         ((uintptr_t)params->address % ucs_get_page_size()))) {
+        ucs_error("UCP_MEM_MAP_FIXED flag requires page aligned address");
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    /* Now, lets check the rest of erroneous cases from the matrix */
+    if (params->address == NULL) {
+        if (!(params->flags & UCP_MEM_MAP_ALLOCATE)) {
+            return UCS_ERR_INVALID_PARAM;
+        }
+    } else if (!(params->flags & UCP_MEM_MAP_ALLOCATE) &&
+               (params->flags & UCP_MEM_MAP_FIXED)) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    return UCS_OK;
+}
+
+static inline int ucp_mem_map_is_allocate(ucp_mem_map_params_t *params)
+{
+    return (params->field_mask & UCP_MEM_MAP_PARAM_FIELD_FLAGS) &&
+           (params->flags & UCP_MEM_MAP_ALLOCATE);
+}
+
+ucs_status_t ucp_mem_map(ucp_context_h context, const ucp_mem_map_params_t *params,
                          ucp_mem_h *memh_p)
 {
-    ucs_status_t status;
-    unsigned uct_flags;
-    ucp_mem_h memh;
+    ucs_status_t            status;
+    ucp_mem_h               memh;
+    ucp_mem_map_params_t    mem_params;
+
+    if (!params) {
```

Comment:
IMHO no need to check for params==NULL since we don't check it in any other function